src/App.tsx

```diff
@@ -1,18 +1,54 @@
-import React from 'react';
+import React, { useEffect, useState } from 'react';
 import './App.css';
 import { getNumbers } from './utils';
+import { Pagination } from './components/Pagination';
 
-// eslint-disable-next-line @typescript-eslint/no-unused-vars
 const items = getNumbers(1, 42)
   .map(n => `Item ${n}`);
 
+const params = new URLSearchParams(window.location.search);
+
 export const App: React.FC = () => {
+  const [
+    perPage,
+    setPerPage,
+  ] = useState(Number(params.get('perPage')) || 5);
+  const [
+    currentPage,
+    setCurrentPage,
+  ] = useState(Number(params.get('page')) || 1);
+  const total = items.length;
+  const firstItemPerPage = currentPage * perPage - perPage + 1;
+  const lastItemPerPage = Math.min(total, currentPage * perPage);
+
+  function changeUrlParams(pageParam: number, perPageParam: number) {
+    params.set('page', String(pageParam));
+    params.set('perPage', String(perPageParam));
+
+    window.history.replaceState({}, '', `${window.location.pathname}?${params}`);
+  }
+
+  function onPageChange(newPage: number) {
+    changeUrlParams(newPage, perPage);
+    setCurrentPage(newPage);
+  }
+
+  function onPerPageChange(newPerPage: number) {
+    changeUrlParams(1, newPerPage);
+    setPerPage(newPerPage);
+    setCurrentPage(1);
+  }
```

Comment:
```suggestion
  function onPageChange(e: React.ChangeEvent<HtmlSelectElement>) {
    const { value } = e.target
    changeUrlParams(1, +value);
    setPerPage(+value);
    setCurrentPage(1);
  }
```
and just pass reference in jsx